README.md
mcmdd
=====

mcmdd monitors, controls, and restarts other application servers. It has
been designed specifically for control of game servers, specifically
Minecraft servers.


debian/changelog
mcmdd (1.0.1-2) stable; urgency=medium

  [ cmastudios ]
  * Add documentation for the user flag
  * Added example config
  * Update paths in systemd script
  * Protocol changes for continouous connections
  * Fedora support
  * Script to make source TARs
  * GitHub documentation

  [ Connor Monahan ]
  * Create README.md

  [ cmastudios ]
  * Update version

  [ Connor Monahan ]
  * Dockerize
  * Give the container java 8 and fix repos

 -- Connor Monahan <admin@cmastudios.me>  Thu, 13 Aug 2015 15:46:03 -0500

mcmdd (0.1-1) unstable; urgency=low

  * Initial release

 -- Connor Monahan <admin@cmastudios.me>  Sat, 27 Sep 2014 14:58:30 -0500


config.c
//
//  config.c
//  mcmdd
//
//  Created by Connor Monahan on 8/16/14.
//  Copyright (c) 2014 Connor Monahan. All rights reserved.
//

#include <stdio.h>
#include <err.h>
#include <stdlib.h>
#include <string.h>

#include "config.h"

struct config_t *config_new(void)
{
    struct config_t *config = malloc(sizeof(struct config_t));
    if (!config)
        err(EXIT_FAILURE, "Failed to allocate memory");
    config->len = 0;
    config->max = 0;
    return config;
}

void config_free(struct config_t *config)
{
    size_t i;
    for (i = 0; i < config->len; ++i) {
        free(config->keys[i]);
        free(config->values[i]);
        free(config->sections[i]);
    }
    free(config->keys);
    free(config->values);
    free(config->sections);
    free(config);
}

static inline void config_add(struct config_t *config, const char *key, const char *value, const char *section)
{
    if (config->len >= config->max) {
        if (config->max < 1)
            config->max = 10;
        config->max *= 2; // double
        config->keys = realloc(config->keys, sizeof(char *) * config->max);
        config->values = realloc(config->values, sizeof(char *) * config->max);
        config->sections = realloc(config->sections, sizeof(char *) * config->max);
        if (!config->keys || !config->values || !config->sections)
            err(EXIT_FAILURE, "Failed to allocate memory");
    }
    config->keys[config->len] = strdup(key);
    config->values[config->len] = strdup(value);
    if (section && section[0]) {
        config->sections[config->len] = strdup(section);
    } else {
        config->sections[config->len] = NULL;
    }
    config->len++;
}

const char *config_get(struct config_t *config, const char *section, const char *key, const char *default_value)
{
    size_t i;
    for (i = 0; i < config->len; ++i) {
        const char *lkey = config->keys[i], *lsec = config->sections[i];
        if (section && !lsec)
            continue;
        if (!section && lsec)
            continue;
        if (section && strcmp(section, lsec) != 0)
            continue;
        if (strcmp(key, lkey) == 0)
            return config->values[i];
    }
    return default_value;
}

enum config_state_t {
    STATE_KEY = 0, // first part of a config option
    STATE_VAL, // second part, the value
    STATE_REM, // comment lines
    STATE_SEC // introduction of a new section
};

int config_load(struct config_t *config, FILE *file)
{
    config->len = 0;
    config->max = 10;
    config->keys = malloc(sizeof(char *) * config->max);
    config->values = malloc(sizeof(char *) * config->max);
    config->sections = malloc(sizeof(char *) * config->max);
    if (!config->keys || !config->values || !config->sections)
        err(EXIT_FAILURE, "Failed to allocate memory");
    char ch;
    int state;
    int pos;
    char key[64], value[1024], section[64];
    state = 0;
    pos = 0;
    section[0] = '\0';
    while ((ch = fgetc(file)) != EOF) {
        if (ch == '\r' || ch == '\n') {
            if (state == STATE_VAL && pos > 0 && strlen(key) > 0) {
                // terminate a value and push it to the config
                value[pos] = '\0';
                config_add(config, key, value, section);
            }
            // reset current state
            state = STATE_KEY;
            pos = 0;
            key[0] = '\0';
            value[0] = '\0';
        } else if (state == STATE_KEY && (ch == '#' || ch == ';')) {
            // comments. checked here because key is at the beginning of the line
            state = STATE_REM;
        } else if (state == STATE_KEY && ch == '=') {
            // at an equals sign, we are done loading the key and need to look for the value
            state = STATE_VAL;
            key[pos] = '\0';
            pos = 0;
        } else if (state == STATE_KEY && (ch == ' ' || ch == '\t')) {
            // ignore spaces in key
        } else if (state == STATE_KEY && ch == '[') {
            state = STATE_SEC;
        } else if (state == STATE_KEY) {
            if (pos > 63)
                return -1;
            key[pos++] = ch;
        } else if (state == STATE_VAL && pos == 0 && (ch == ' ' || ch == '\t')) {
            // ignore space at start of value
        } else if (state == STATE_VAL) {
            if (pos > 1023)
                return -1;
            value[pos++] = ch;
        } else if (state == STATE_SEC && ch == ']') {
            // terminate the section. note it is not reset, as it will stay the section until there is a new section
            // it is set to comment mode so everything else is ignored
            section[pos] = '\0';
            state = STATE_REM;
            pos = 0;
            key[0] = '\0';
            value[0] = '\0';
        } else if (state == STATE_SEC) {
            if (pos > 63)
                return -1;
            section[pos++] = ch;
        }
    }
    return 0;
}


server.c
//
//  server.c
//  mcmdd
//
//  Created by Connor Monahan on 8/16/14.
//  Copyright (c) 2014 Connor Monahan. All rights reserved.
//

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>
#include <err.h>
#include <stdlib.h>
#include <signal.h>
#include <assert.h>

#include "server.h"

struct server_t *server_new(const char *path, const char *command, const char *id)
{
    struct server_t *server = malloc(sizeof(struct server_t));
    server->id = strdup(id);
    server->path = strdup(path);
    char **argv = malloc(sizeof(char *) * strlen(command));

	size_t len = 0;
	char *str_ = malloc(strlen(command) + 1);
	size_t i, tokpos;
	for (i = 0, tokpos = 0; i < strlen(command); i++, tokpos++) {
		if (command[i] == '\\') {
			str_[tokpos] = command[++i];
		} else if (command[i] == ' ') {
			if (tokpos == 0) {
				tokpos = -1;
				continue;
			}
            str_[tokpos] = '\0';
            argv[len++] = strdup(str_);
			tokpos = -1;
		} else {
			str_[tokpos] = command[i];
        }
	}
    if (tokpos > 0) {
        str_[tokpos] = '\0';
        argv[len++] = strdup(str_);
    }
    argv[len] = NULL; // must be null terminated for execv
    server->argv = argv;
    server->linsp = 0;
    bzero(server->lines, SERVER_MAXLINES);
	free(str_);
    return server;
}

static inline void server_cleanup(struct server_t *server)
{
    size_t i;
    char *str;
    
    i = 0;
    str = server->lines[i++];
    while (str) {
        free(str);
        str = server->lines[i++];
    }
}

void server_free(struct server_t *server)
{
    size_t i;
    char *str;
    free(server->id);
    free(server->path);
    i = 0;
    str = server->argv[i++];
    while (str) {
        free(str);
        str = server->argv[i++];
    }
    server_cleanup(server);
    free(server);
}

static void add_line(struct server_t *server, const char *line)
{
    if (server->linsp >= SERVER_MAXLINES)
        server->linsp = 0;
    if (server->lines[server->linsp])
        free(server->lines[server->linsp]);
    server->lines[server->linsp++] = strdup(line);
}

static inline void process_line(struct server_t *server, const char *line)
{
    add_line(server, line);
    printf("[%s] #%2d: %s\n", server->id, server->linsp, line);
    if (server->status == STATUS_STARTING && strstr(line, "Done"))
        server->status = STATUS_RUNNING;
    time(&server->last_read);
}

static void read_line(int fd, struct server_t *server)
{
    char ch, buf[SERVER_LINEMAX];
    int sp;
    
    sp = 0;
    while (read(fd, &ch, 1) != 0) {
        // lines longer than max are cycled around
        if (ch == '\n' || sp >= SERVER_LINEMAX) {
            // null-terminate
            buf[sp++] = '\0';
            process_line(server, buf);
            sp = 0;
        } else {
            buf[sp++] = ch;
        }
    }
}

int server_send(struct server_t *server, const char *message)
{
    if (server->status == STATUS_STOPPED)
        return -1;
    add_line(server, message);
    printf("[%s] < %s\n", server->id, message);
    write(server->pipein, message, strlen(message));
    return 0;
}

void server_stop(struct server_t *server, int exit)
{
    if (exit == EXIT_FULL)
        server->ctrl = CTRL_EXIT;
    else if (exit == EXIT_PAUSE)
        server->ctrl = CTRL_PAUSE;
    if (server->status == STATUS_STOPPED)
        return;
    server->status = STATUS_STOPPING;
    server_send(server, "stop\n");
}

int server_kill(struct server_t *server, int exit)
{
    if (exit == EXIT_FULL)
        server->ctrl = CTRL_EXIT;
    else if (exit == EXIT_PAUSE)
        server->ctrl = CTRL_PAUSE;
    if (server->status == STATUS_STOPPED)
        return -1;
    printf("[%s] Killing server process %d\n", server->id, server->pid);
    add_line(server, "Server process killed");
    return kill(server->pid, SIGKILL);
}

void server_resume(struct server_t *server)
{
    server->ctrl = CTRL_LAUNCH;
}

int server_start(struct server_t *server)
{
    pid_t cpid, pid;
    int pipeout[2], pipein[2];
    // pipe for reading from the server console
    pipe(pipeout);
    // pipe for sending commands and messages to the server
    pipe(pipein);
    // status field is to allow other threads to check how the server is doing
    server->status = STATUS_STARTING;
    server->pipein = pipein[1];
    time(&server->start);
    cpid = fork(); // program execution flow 'forks' here
    if (cpid == 0) {
        // child process started for server
        // assign the standard in, out, and error so the server can be monitored
        dup2(pipein[0], 0);
        dup2(pipeout[1], 1);
        dup2(pipeout[1], 2);
        // close the original pipes (the functions above created clones)
        close(pipeout[0]);
        close(pipeout[1]);
        close(pipein[1]);
        close(pipeout[1]);
        // go to the server's data directory
        chdir(server->path);
        // launch it, calling java (or whatever) from the server's PATH
        execvp(server->argv[0], server->argv);
        err(1, "server exec failed for %s", server->id);
    } else if (cpid > 0) {
        // parent process to monitor server
        int status;
        // close the ends of the pipe that we shouldn't be using
        // e.g. writing to the pipe that returns the child's output
        close(pipeout[1]);
        close(pipein[0]);
        server->pid = cpid;
        printf("[%s] Starting on PID %d.\n", server->id, cpid);
        // read until the server stops
        read_line(pipeout[0], server);
        // close the pipes as we are all done
        close(pipeout[0]);
        close(pipein[1]);
        // get the status and prevent creating zombies
        pid = wait(&status);
        server->status = STATUS_STOPPED;
        printf("[%s] PID %d exists with %d.\n", server->id, pid, status);
        return status;
    } else {
        err(1, "server fork failed for %s", server->id);
    }
}

server.h
//
//  server.h
//  mcmdd
//
//  Created by Connor Monahan on 8/16/14.
//  Copyright (c) 2014 Connor Monahan. All rights reserved.
//

#ifndef mcmdd_server_h
#define mcmdd_server_h

#include <time.h>
#include <sys/types.h>

enum server_status_t {
    // server not running
    STATUS_STOPPED = 0,
    // listener has not received the "Done" message yet from minecraft
    STATUS_STARTING,
    // server has reported a successful start
    STATUS_RUNNING,
    // server_stop was used
    STATUS_STOPPING,
};

enum server_control_t {
    // internal
    CTRL_CLEAN = 0,
    // instructs restarter to quit completely
    CTRL_EXIT,
    // alerts a restarter listening to start again
    CTRL_LAUNCH,
    // causes the restarter to pause until receiving a launch
    CTRL_PAUSE
};

enum exit {
    EXIT_PAUSE = 0,
    EXIT_FULL = 1,
    EXIT_RESTART = 2
};

#define SERVER_MAXLINES 1024
#define SERVER_LINEMAX 1024

struct server_t {
    pid_t pid;
    char *path;
    char *const *argv;
    char *id;
    enum server_status_t status;
    enum server_control_t ctrl;
    int pipein, linsp;
    char *lines[SERVER_MAXLINES];
    time_t start, last_read;
};

struct server_t *server_new(const char *path, const char *command, const char *id);
void server_free(struct server_t *server);
int server_start(struct server_t *server);
int server_send(struct server_t *server, const char *message);
void server_stop(struct server_t *server, int exit);
int server_kill(struct server_t *server, int exit);
void server_resume(struct server_t *server);

#endif


control.c
//
//  control.c
//  mcmdd
//
//  Created by Connor Monahan on 8/17/14.
//  Copyright (c) 2014 Connor Monahan. All rights reserved.
//

#include <stdio.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <poll.h>
//#include <sys/un.h>
#include <err.h>
#include <string.h>
#include <unistd.h>
#include <pthread.h>
#include <stdlib.h>
#include <errno.h>
#include "config.h"
#include "server.h"

static struct sockaddr_in sin;
//static struct sockaddr_un sun;
static int listener;
extern struct config_t *config;

struct server_t *get_server(const char *id);

void control_init()
{
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = INADDR_ANY;
    sin.sin_port = htons(atoi(config_get(config, NULL, "port", "8361")));
    
//    sun.sun_family = AF_UNIX;
//    strcpy(sun.sun_path, "mcmdd.sock");
//    unlink(sun.sun_path);
    
    listener = socket(AF_INET, SOCK_STREAM, 0);
    
    if (bind(listener, (struct sockaddr *) &sin, sizeof(sin)) < 0)
        err(1, "bind to IP socket");
    
//    if (bind(listener, (struct sockaddr *) &sun, sizeof(sin)) < 0)
//        err(1, "bind to Unix socket");
//    
    if (listen(listener, 5) < 0)
        err(1, "listen");
}

/*!
 * @return number of bytes read, or status. Returns -1 on overflow
 *   or 0 on end of file.
 */
static int read_line(char *out, int fd, int max)
{
    int bytes, status;
    char ch;
    struct pollfd pfd;
    
    pfd.fd = fd;
    pfd.events = POLLIN;
    bytes = 0;
    status = 1;
    while (status > 0) {
        status = poll(&pfd, 1, 10000);
        if (status == 0)
            return -2;
        else if (status == -1 && errno == EINTR)
            continue; // interrupted sys call
        else if (status == -1)
            err(1, "read_line poll");
        status = recv(fd, &ch, 1, 0);
        if (status == 0)
            return -1;
        else if (status == -1)
            err(1, "read_line recv");
        
        if (bytes >= max)
            return -1;
        if (ch == '\n') {
            out[bytes] = '\0';
            return bytes;
        } else {
            out[bytes++] = ch;
        }
    }
    return 0;
}

int valid(const char *key, const char *server)
{
    if (!key || !server || strlen(key) < 1 || strlen(server) < 1)
        return 0;
    if (strstr(config_get(config, NULL, "servers", ""), server) == NULL)
        // Nonexistant server
        return 0;
    char *token, *string, *tofree;
    int valid;
    
    tofree = string = strdup(config_get(config, server, "auth", config_get(config, NULL, "auth", "")));
    if (!string)
        err(1, "strdup");
    
    valid = 0;
    while ((token = strsep(&string, " ")) != NULL)
        if (strcmp(key, token) == 0)
            valid = 1;
    
    free(tofree);
    return valid;
}

#define APPNAME "mcmdd/1.0.1\n"
#define INVALID "ERR Invalid command.\n"
#define SVNEXT "OK Need key.\n"
#define KYNEXT "OK Need server.\n"
#define BADKEY "ERR Bad login.\n"
#define OKKEY "OK Logged in.\n"
#define INTERR "ERR Internal error.\n"
#define OKEXEC "OK Command sent.\n"
#define STATF "OK Stats %d %.f\n"
#define EOFF "ERR Server is off.\n"
#define TSTART "OK Send start.\n"
#define TEND "OK Send end.\n"

static inline void qwrite(int fd, const char *message)
{
    write(fd, message, strlen(message));
}

static inline void send_log(int fd, struct server_t *server, const char *start_line)
{
    int sp, bufsp, found_line;
    char buf[(SERVER_LINEMAX + 2) * SERVER_MAXLINES];

    bufsp = 0;
    found_line = 0;
    qwrite(fd, TSTART);
    if (server->lines[server->linsp]) {
        // means data has been overwritten, so start reading from linsp forward
        for (sp = server->linsp; sp < SERVER_MAXLINES; ++sp) {
            // add line to buffer at current position
            sprintf(buf + bufsp, "%s\n", server->lines[sp]);
            // buffer is written into like one giant block
            bufsp += strlen(server->lines[sp]) + 1; // +1 to overwrite the last \0
            if (start_line && !found_line && strstr(start_line, server->lines[sp]) == start_line) {
                found_line = 1;
                bufsp = 0;
            }
        }
    }
    for (sp = 0; sp < server->linsp; ++sp) {
        sprintf(buf + bufsp, "%s\n", server->lines[sp]);
        bufsp += strlen(server->lines[sp]) + 1;
        if (start_line && !found_line && strstr(start_line, server->lines[sp]) == start_line) {
            found_line = 1;
            bufsp = 0;
        }
    }
    // send the whole buffer
    write(fd, buf, bufsp);
    qwrite(fd, TEND);
}

void control_read(int fd)
{
    qwrite(fd, APPNAME);
    char tmp[256], msg[256];
    char *key, *server;
    int vald;
    int ka;
    int r;
    struct server_t *serv;
    
    key = server = NULL;
    vald = 0;
    serv = NULL;
    ka = 0;
    while (1) {
        r = read_line(tmp, fd, 256);
        if (r == -2 && ka)
            continue;
        else if (r < 1)
            goto clean;
        if (strstr(tmp, "SERVER ") == tmp) {
            free(server);
            server = strdup(tmp + 7);
            vald = valid(key, server);
            if (vald)
                qwrite(fd, OKKEY);
            else if (key)
                qwrite(fd, BADKEY);
            else
                // purposely doesn't notify for invalid server, for security
                qwrite(fd, SVNEXT);
        } else if (strstr(tmp, "KEY ") == tmp) {
            free(key);
            key = strdup(tmp + 4);
            vald = valid(key, server);
            if (vald)
                qwrite(fd, OKKEY);
            else if (key)
                qwrite(fd, BADKEY);
            else
                qwrite(fd, KYNEXT);
        } else if (strstr(tmp, "EXEC ") == tmp) {
            if (!serv) {
                qwrite(fd, BADKEY);
                continue;
            }
            sprintf(msg, "%s\n", tmp + 5);
            if (server_send(serv, msg) == 0)
                qwrite(fd, OKEXEC);
            else
                qwrite(fd, EOFF);
        } else if (strstr(tmp, "KILL") == tmp) {
            if (!serv) {
                qwrite(fd, BADKEY);
                continue;
            }
            if (server_kill(serv, EXIT_PAUSE) == 0)
                qwrite(fd, OKEXEC);
            else
                qwrite(fd, INTERR);
        } else if (strstr(tmp, "STOP") == tmp) {
            if (!serv) {
                qwrite(fd, BADKEY);
                continue;
            }
            server_stop(serv, EXIT_PAUSE);
            qwrite(fd, OKEXEC);
        } else if (strstr(tmp, "RESTART") == tmp) {
            if (!serv) {
                qwrite(fd, BADKEY);
                continue;
            }
            server_stop(serv, EXIT_RESTART);
            qwrite(fd, OKEXEC);
        } else if (strstr(tmp, "START") == tmp) {
            if (!serv) {
                qwrite(fd, BADKEY);
                continue;
            }
            server_resume(serv);
            qwrite(fd, OKEXEC);
        } else if (strstr(tmp, "STATUS") == tmp) {
            if (!serv) {
                qwrite(fd, BADKEY);
                continue;
            }
            sprintf(msg, STATF, serv->status, difftime(time(NULL), serv->start));
            qwrite(fd, msg);
        } else if (strstr(tmp, "LOG") == tmp) {
            if (!serv) {
                qwrite(fd, BADKEY);
                continue;
            }
            if (tmp[3] == ' ') {
                send_log(fd, serv, tmp + 4);
            } else {
                send_log(fd, serv, NULL);
            }
        } else if (strstr(tmp, "KEEPALIVE") == tmp) {
            ka = 1;
        } else {
            qwrite(fd, INVALID);
        }
        if (vald) {
            serv = get_server(server);
            if (!serv) {
                qwrite(fd, INTERR);
                goto clean;
            }
        } else {
            serv = NULL;
        }
    }
clean:
    close(fd);
    free(key);
    free(server);
}

void *control_thread(void *data)
{
    control_read(*((int *)data));
    pthread_exit(NULL);
}

void control_accept()
{
    int fd, rc;
    void *status;
    struct sockaddr_in ss;
    socklen_t slen = sizeof(ss);
    while (1) {
        fd = accept(listener, (struct sockaddr *) &ss, &slen);
        if (fd == -1 && errno == EINTR)
            continue; // interrupted system call
        else if (fd == -1)
            err(1, "accept");
        pthread_t threadId;
        rc = pthread_create(&threadId, NULL, &control_thread, (void *) &fd);
        if (pthread_detach(threadId) != 0) {
            warn("Failed to detach control thread, running sync");
            pthread_join(threadId, &status);
        }
    }
}

void control_stop()
{
    shutdown(listener, SHUT_RDWR);
    close(listener);
}


main.c
//
//  main.c
//  mcmdd
//
//  Created by Connor Monahan on 8/15/14.
//  Copyright (c) 2014 Connor Monahan. All rights reserved.
//

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <err.h>
#include <sys/stat.h>
#include <pthread.h>
#include <sys/types.h>
#include <pwd.h>

#include "config.h"
#include "server.h"

const char *program_name;
struct config_t *config;
struct server_t **servers;
pthread_t **threads;
int servers_sp, threads_sp;

void control_init();
void control_accept();
void control_stop();

static void usage(void)
{
    fprintf(stderr, "usage: %s [-nf] [-d path]\n", program_name);
    exit(1);
}

static void load_config()
{
    config = config_new();
    FILE *config_file = fopen("mcmdd.conf", "r");
    if (!config_file)
        err(EXIT_FAILURE, "Failed to load config");
    config_load(config, config_file);
    fclose(config_file);
}

struct server_t *get_server(const char *id)
{
    size_t i;
    for (i = 0; i < servers_sp; ++i)
        if (strcmp(id, servers[i]->id) == 0)
            return servers[i];
    return NULL;
}

static void load_server(const char *name)
{
    const char *path, *command;

    printf("[%s] Loading\n", name);
    
    path = config_get(config, name, "path", "");
    command = config_get(config, name, "command", DEFAULT_COMMAND);
    
    servers[servers_sp++] = server_new(path, command, name);
}

static void load_servers()
{
    char *svr, *string, *tofree;
    const char *svrs;
    
    svrs = config_get(config, NULL, "servers", NULL);
    if (!svrs)
        errx(1, "No servers list found in config");
    
    servers_sp = 0;
    servers = malloc(sizeof(struct server_t *) * strlen(svrs));
    tofree = string = strdup(svrs);
    if (!string)
        err(1, "strdup failed");
    
    while ((svr = strsep(&string, " ")) != NULL)
        load_server(svr);
    
    free(tofree);
    servers = realloc(servers, sizeof(struct server_t *) * servers_sp);
}

/*!
 * @return 0 for exit immediate, 1 for resume
 */
static inline int server_pause_loop(struct server_t *server)
{
    while (1) {
        if (server->ctrl == CTRL_EXIT)
            return 0;
        // resume the server, called by control
        if (server->ctrl == CTRL_LAUNCH)
            return 1;
        sleep(1);
    }
}

void *thread_start_wrapper(void *ptr)
{
    struct server_t *server = ptr;
    while (1) {
        server->ctrl = CTRL_CLEAN;
        server_start(server);
        // set when the daemon is killing all other servers, so quit
        if (server->ctrl == CTRL_EXIT)
            return NULL;
        // set by control when a shutdown is anticipated
        if (server->ctrl == CTRL_PAUSE
            && server_pause_loop(server) == 0)
            return NULL;
    }
}

static void run_server(struct server_t *server, int id)
{
    pthread_t *thread;
    int rc;
    
    thread = malloc(sizeof(pthread_t *));
    rc = pthread_create(thread, NULL, thread_start_wrapper, server);
    if (rc)
        err(1, "pthread_create for %s", server->id);
    threads[threads_sp] = thread;
}

static void run_servers()
{
    size_t i;
    
    threads = malloc(sizeof(pthread_t *) * servers_sp);
    for (i = 0, threads_sp = 0; i < servers_sp; ++i, ++threads_sp)
        run_server(servers[i], i);
}

static void kill_servers()
{
    size_t i;
    void *status;
    
    puts("[daemon] Killing all servers");
    for (i = 0; i < servers_sp; ++i) {
        server_kill(servers[i], EXIT_FULL);
        pthread_join(*threads[i], &status);
    }
}

static void stop_servers()
{
    size_t i;
    void *status;
    
    puts("[daemon] Stopping all servers");
    for (i = 0; i < servers_sp; ++i) {
        server_stop(servers[i], EXIT_FULL);
        pthread_join(*threads[i], &status);
    }
}

static void cleanup()
{
    size_t i;

    // these are reassigned to files if forked
    fclose(stdout);
    fclose(stderr);
    
    config_free(config);
    for (i = 0; i < servers_sp; ++i) {
        server_free(servers[i]);
    }
    for (i = 0; i < threads_sp; ++i) {
        free(threads[i]);
    }
    free(servers);
    free(threads);
}

void signal_handler(int signum)
{
    // ignore notices about terminating child processes
    if (signum == SIGCHLD)
        return;
    
    printf("[daemon] Caught signal %d.\n", signum);
    // SIGINT is regarded as a safe shutdown signal
    if (signum == SIGINT)
        stop_servers();
    // SIGTERM is regarded as an emergency shutdown maneuver
    // such as at shutdown, when we receive SIGTERM and then only have
    // a few seconds before we are killed
    if (signum == SIGTERM)
        kill_servers();
    control_stop();
    puts("[daemon] Cleaning up");
    cleanup();
    exit(1);
}

static inline void fork_background()
{
    pid_t pid, sid;

    pid = fork();
    if (pid < 0)
        err(EXIT_FAILURE, "Failed to fork daemon to background");
    else if (pid > 0)
        // Is parent, so exit
        exit(EXIT_SUCCESS);
    // create files with mode 666
    umask(0);
    // turn the child process into a normal process
    sid = setsid();
    if (sid < 0)
        // failed to assert independence :)
        exit(EXIT_FAILURE);
    close(STDIN_FILENO);
    close(STDOUT_FILENO);
    close(STDERR_FILENO);
    freopen("mcmdd.log", "a", stdout);
    freopen("mcmdd.err", "a", stderr);
    fchmod(0, S_IRUSR | S_IWUSR);
    fchmod(1, S_IRUSR | S_IWUSR);
    setbuf(stdout, NULL);
    setbuf(stderr, NULL);
    FILE *pidfile = fopen("mcmdd.pid", "w");
    if (!pidfile)
        err(1, "Failed to write pid file");
    fprintf(pidfile, "%d\n", sid);
    fclose(pidfile);
    chmod("mcmdd.pid", S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
}

static inline void change_user(const char *name)
{
    struct passwd *user = getpwnam(name);
    if (!user)
        errx(1, "User %s not found", name);
    if (setgid(user->pw_gid) < 0)
        err(1, "Changing group");
    if (setuid(user->pw_uid) < 0)
        err(1, "Changing user");
}

int main(int argc, char **argv)
{
    int ch, dofork;
    char *data_dir;
    
    program_name = argv[0];
    dofork = 0; // change to 1 for release
    data_dir = NULL;
    
    while ((ch = getopt(argc, argv, "nfd:u:")) != -1) {
        switch (ch) {
            case 'n':
                dofork = 0;
                break;
            case 'f':
                dofork = 1;
                break;
            case 'd':
                data_dir = strdup(optarg);
                break;
            case 'u':
                change_user(optarg);
                break;
            case '?':
            default:
                usage();
        }
    }
    argc -= optind;
    argv += optind;
    
    if (data_dir && chdir(data_dir) < 0)
        err(EXIT_FAILURE, "Failed to open data dir %s", data_dir);
    free(data_dir);

    if (dofork)
        fork_background();
    
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    signal(SIGCHLD, signal_handler);
    
    load_config();
    control_init();
    load_servers();
    run_servers();
    control_accept();
}
